services/src/main/java/org/keycloak/authentication/requiredactions/GrantManagement.java

```diff
@@ -0,0 +1,248 @@
+        }
+
+        Set<String> grantScopes = new HashSet<String>();
+        boolean updateConsentRequired = false;
+        for (String clientScopeId : authSession.getClientScopes()) {
+            ClientScopeModel clientScope = KeycloakModelUtils.findClientScopeById(realm, client, clientScopeId);
+            if (clientScope != null) {
+                if (!grantedConsent.isClientScopeGranted(clientScope) && clientScope.isDisplayOnConsentScreen()) {
+                    grantedConsent.addGrantedClientScope(clientScope);
+                    updateConsentRequired = true;
+                }
+
+                if (grantedConsent.isClientScopeGranted(clientScope)) {
+                    grantScopes.add(clientScope.getName());
+                }
+            } else {
+                logger.warnf("Client scope or client with ID '%s' not found", clientScopeId);
+            }
+        }
+
+        if (updateConsentRequired) {
+            context.getSession().users().updateConsent(realm, user.getId(), grantedConsent);
+        }
+
+        if (Constants.GRANT_MANAGEMENT_ACTION_CREATE.equals(grantManagementAction)) {
+            try {
+
+                userGrantModel = new UserGrantModel();
+                grantId = Base64Url.encode(KeycloakModelUtils.generateSecret());
+                userGrantModel.setGrantId(grantId);
+                userGrantModel.setClientId(client.getClientId());
+                userGrantModel.setUserId(user.getId());
+                userGrantModel.setScopes(grantScopes);
+                userGrantModel.setClaims(authSession.getAuthNote(OIDCLoginProtocol.CLAIMS_PARAM));
+                //TODO: when merging with RAR, userGrantModel.setAuthorizationDetails(rarProcessor.finaliseAuthorizationDetails(grantManagementAction, formData, authSession.getAuthNote(OIDCLoginProtocol.AUTHORIZATION_DETAILS_PARAM)));
+                userGrantModel.setAuthorizationDetails(authSession.getAuthNote(OIDCLoginProtocol.AUTHORIZATION_DETAILS_PARAM));
+                userGrantModel.setCreatedDate(currentTime);
+                userGrantModel.setLastUpdatedDate(currentTime);
+                grantManagementProvider.adduserGrant(realm, userGrantModel);
+
+            } catch (Exception e) {
+                cleanSession(context, RequiredActionContext.KcActionStatus.ERROR);
+                context.failure();
+                event.error(Errors.INVALID_REQUEST);
+                return;
+            }
+        }
+
+        if (Constants.GRANT_MANAGEMENT_ACTION_REPLACE.equals(grantManagementAction)) {
+
+            try {
+                userGrantModel.setScopes(grantScopes);
+                userGrantModel.setClaims(authSession.getAuthNote(OIDCLoginProtocol.CLAIMS_PARAM));
+                //TODO: when merging with RAR, userGrantModel.setAuthorizationDetails(rarProcessor.finaliseAuthorizationDetails(grantManagementAction, formData, authSession.getAuthNote(OIDCLoginProtocol.AUTHORIZATION_DETAILS_PARAM)));
```

Comment:
@tnorimat How can we deal here, because grant management needs RAR to complete.

Reply:
If supporting RAR at first is acceptable, we can resume grant management support after RAR is supported.
If not, could we skip this point? It means that grant management is partially supported at the first time.

Reply:
So, It seems that RAR is supported at first as you've working on https://github.com/tnorimat/keycloak/pull/24 .
Could you treat your RAR support at first before this grant management support?

Reply:
understand.